File: web/lib/config/ports.ts
/**
 * Centralized port configuration for the Pinchy application.
 *
 * These ports are used across the frontend and must stay synchronized
 * with the backend configuration.
 */

// Type-safe port configuration
const DEFAULT_GATEWAY_PORT = 3131;
const DEFAULT_FRONTEND_DEV_PORT = 3000;
const DEFAULT_VITE_PORT = 5173;
const DEFAULT_ANTHROPIC_COMPAT_PORT = 8080;
const DEFAULT_OLLAMA_PORT = 11434;

/**
 * Port numbers used throughout the application.
 * @readonly
 */
export const PORTS = {
  /** Backend gateway port - can be overridden via env var */
  GATEWAY: Number(process.env.NEXT_PUBLIC_GATEWAY_PORT) || DEFAULT_GATEWAY_PORT,
  /** Frontend dev server port (Next.js) */
  FRONTEND_DEV: DEFAULT_FRONTEND_DEV_PORT,
  /** Vite dev server port */
  VITE: DEFAULT_VITE_PORT,
  /** Anthropic/OpenAI compatible API port */
  ANTHROPIC_COMPAT: DEFAULT_ANTHROPIC_COMPAT_PORT,
  /** Ollama local API port */
  OLLAMA_DEFAULT: DEFAULT_OLLAMA_PORT,
} as const;

/**
 * Get the WebSocket URL for the main gateway WebSocket.
 * Handles development vs production environments.
 */
export function getGatewayWebSocketUrl(): string {
  if (typeof window === 'undefined') {
    return `ws://127.0.0.1:${PORTS.GATEWAY}/ws`;
  }

  const isDev = process.env.NODE_ENV === 'development';
  const isLocalhost = window.location.hostname === 'localhost' || window.location.hostname === '127.0.0.1';
  const protocol = window.location.protocol === 'https:' ? 'wss' : 'ws';

  if (isDev && isLocalhost) {
    return `ws://127.0.0.1:${PORTS.GATEWAY}/ws`;
  }

  if (isDev) {
    return `${protocol}://${window.location.hostname}:${PORTS.GATEWAY}/ws`;
  }

  // Production: use same-origin
  return `${protocol}://${window.location.host}/ws`;
}

/**
 * Get the WebSocket URL for the logs WebSocket endpoint.
 */
export function getLogsWebSocketUrl(): string {
  if (typeof window === 'undefined') {
    return `ws://127.0.0.1:${PORTS.GATEWAY}/ws/logs`;
  }

  const protocol = window.location.protocol === 'https:' ? 'wss' : 'ws';
  return `${protocol}://${window.location.hostname}:${PORTS.GATEWAY}/ws/logs`;
}

/**
 * Get the API base URL for backend requests.
 * Used primarily in development mode with Next.js rewrites.
 */
export function getApiBaseUrl(): string {
  return `http://127.0.0.1:${PORTS.GATEWAY}/api`;
}

File: src/ports.rs
//! Centralized port configuration for the Pinchy application.
//!
//! These ports are used across the backend and must stay synchronized
//! with the frontend configuration in `web/lib/config/ports.ts`.

/// Backend gateway port (default: 3131)
/// Can be overridden via the `PINCHY_GATEWAY_ADDR` environment variable
pub const GATEWAY_DEFAULT: u16 = 3131;

/// Frontend development port (Next.js default)
pub const FRONTEND_DEV: u16 = 3000;

/// Vite development port
pub const VITE_DEV: u16 = 5173;

/// Anthropic/OpenAI compatible API port for local providers
pub const ANTHROPIC_COMPAT: u16 = 8080;

/// Default Ollama API port
pub const OLLAMA_DEFAULT: u16 = 11434;

/// LM Studio default port
pub const LMSTUDIO_DEFAULT: u16 = 1234;

/// vLLM default port
pub const VLLM_DEFAULT: u16 = 8000;

/// CORS allowed ports for development
pub const CORS_ALLOWED_PORTS: &[u16] = &[GATEWAY_DEFAULT, FRONTEND_DEV, VITE_DEV, ANTHROPIC_COMPAT];

/// Default gateway bind address
pub const GATEWAY_BIND_DEFAULT: &str = "0.0.0.0:3131";

/// Default gateway address for local connections (used by CLI status check)
pub const GATEWAY_LOCAL_DEFAULT: &str = "127.0.0.1:3131";
